File: app/socketevents.py
from app import socketio, db
from app.model import LobbyPlayers
from flask import session
from flask_socketio import SocketIO, emit, join_room, leave_room
from flask_login import current_user

@socketio.on("connect")
def connect():
    if not "lobby_id" in session:
        socketio.stop()

    lobby_id = session["lobby_id"]
    join_room(lobby_id)


@socketio.on("disconnect")
def disconnect():
    lobby_id = session["lobby_id"]
    leave_room(lobby_id)
    session.pop("lobby_id")


# Events all users can send
@socketio.on("player_text")
def player_text(data):
    """Broadcasts the player's username and text message.

    data is in the format:
        data = {
            "body": "text_message"
        }

    Sent data is in the format:
        data_to_send = {
            "sender_username": "sender_username", 
            "body": "text_message"
        }
    """
    lobby_id = session["lobby_id"]
    text_message = data["body"]

    # TODO: Add message to database if needed

    sender_username = current_user.Username
    data_to_send = {
        "sender_username": sender_username, 
        "body": text_message
    }
    emit("player_text", data_to_send, to=lobby_id)


# Events only the host can send
@socketio.on("add_tag")
def add_tag(data):
    """Broadcasts the tag that was added.

    data is in the format:
        data = {
            "body": "tag"
        }

    Sent data is in the format:
        data_to_send = {
            "body": "tag"
        }
    """
    lobby_id = session["lobby_id"]
    if not is_lobby_host(current_user.get_id(), lobby_id): 
        return
    tag = data["body"]
    # TODO: Add tag to database
    data_to_send = {
        "body": tag
    }
    emit("add_tag", data_to_send, to=lobby_id)

@socketio.on("remove_tag")
def remove_tag(data):
    """Broadcasts the tag that was removed.

    data is in the format:
        data = {
            "body": "tag"
        }

    Sent data is in the format:
        data_to_send = {
            "body": "tag"
        }
    """
    lobby_id = session["lobby_id"]
    if not is_lobby_host(current_user.get_id(), lobby_id): 
        return
    tag = data["body"]
    # TODO: Remove tag from database
    data_to_send = {
        "body": tag
    }
    emit("remove_tag", data_to_send, to=lobby_id)

@socketio.on("change_lobby_name")
def change_lobby_name(data):
    """Broadcasts the new lobby name.

    data is in the format:
        data = {
            "body": "new lobby name"
        }

    Sent data is in the format:
        data_to_send = {
            "body": "new lobby name"
        }
    """
    lobby_id = session["lobby_id"]
    if not is_lobby_host(current_user.get_id(), lobby_id): 
        return
    new_lobby_name = data["body"]
    # TODO: Update lobby name in database
    data_to_send = {
        "body": new_lobby_name
    }
    emit("change_lobby_name", data_to_send, to=lobby_id)

@socketio.on("change_description")
def change_description(data):
    """Broadcasts the new description.

    data is in the format:
        data = {
            "body": "new description"
        }

    Sent data is in the format:
        data_to_send = {
            "body": "new description"
        }
    """
    lobby_id = session["lobby_id"]
    if not is_lobby_host(current_user.get_id(), lobby_id): 
        return
    new_description = data["body"]
    # TODO: Update lobby description in database
    data_to_send = {
        "body": new_description
    }
    emit("change_lobby_name", data_to_send, to=lobby_id)

@socketio.on("change_time_schedule")
def change_time_schedule(data):
    """Broadcasts the new time schedule.

    data is in the format:
        data = {
            "body": "new time schedule"
        }

    Sent data is in the format:
        data_to_send = {
            "body": "new time schedule"
        }
    """
    lobby_id = session["lobby_id"]
    if not is_lobby_host(current_user.get_id(), lobby_id): 
        return
    new_time_schedule = data["body"]
    # TODO: Update lobby time schedule in database
    data_to_send = {
        "body": new_time_schedule
    }
    emit("change_time_schedule", data_to_send, to=lobby_id)

def is_lobby_host(user_id, lobby_id):
    """Checks if user_id is the host of lobby_id."""
    # TODO
    return True

File: app/static/scripts/socket-events.js
$(document).ready(function() {
  let url = window.location.hostname + ":" + window.location.port; //"http://localhost:5000";
  let socket = io.connect(url);

  // Events for players joining or leaving
  socket.on("player_join", function(data) {
    let username = data.sender_username;

    $("#players").append("<span class='lobby-player'>" + username + " </span>");
    let text = "<div class='chat-text'><span class='server-text'>" +
                "<span class='username'>" + username + 
                "</span> has joined</span></div>";
    $("#chat-log").append(text);

    let currPlayerCount = Number($("#current-num-players").text());
    currPlayerCount += 1;
    $("#current-num-players").text(currPlayerCount); 
  });

  socket.on("player_leave", function(data) {
    let username = data.sender_username;

    $(".lobby-player").map(function() {
      if (this.innerHTML === (username + " ")) this.remove();
    });
    let text = "<div class='chat-text'><span class='server-text'>" +
                "<span class='username'>" + username + 
                "</span> has left</span></div>";
    $("#chat-log").append(text);

    let currPlayerCount = Number($("#current-num-players").text());
    currPlayerCount -= 1;
    $("#current-num-players").text(currPlayerCount);
  });

  // Events all users can send
  socket.on("player_text", function(data) {
    let username = data.sender_username;
    let playerText = data.body;

    let text = "<div class='chat-text'><span class='username'>" +
                username + "</span>: " +
                playerText + "</div>";
    $("#chat-log").append(text);
  });

  // Events only the host can send
  socket.on("add_tag", function(data) {
    let tag = data.body;

    $("#tags").append("<span class='tag'>" + tag + "</span>");
  });

  socket.on("remove_tag", function(data) {
    let tag = data.body;
    $(".tag").map(function() {
      if (this.innerHTML === tag) this.remove();
    });
  });

  socket.on("change_lobby_name", function(data) {
    let newLobbyName = data.body;

    $("#lobby-name").html(newLobbyName);
  });

  socket.on("change_description", function(data) {
    let newDescription = data.body;

    $("#lobby-description").html(newDescription);
  });

  socket.on("change_time_schedule", function(data) {
    // TODO: change schedule, also need to decide on data.body format,
    //       or maybe data.body is just html
  });

  // Events that the user sends to the server
  $("#chat-input").keypress(function(e) {
    let keyCode = e.keyCode || e.which;
    if (keyCode === 13) { // "Enter" key code
      let text = $("#chat-input").val();
      $("#chat-input").val("");
      let dataToSend = {
        body: text
      };
      socket.emit("player_text", dataToSend);
    }
  });
});
